pkg/compose/secrets.go

```diff
@@ -94,47 +114,135 @@ func (s *composeService) injectConfigs(ctx context.Context, project *types.Proje
 	return nil
 }
 
-func createTar(env string, config types.FileReferenceConfig) (bytes.Buffer, error) {
-	value := []byte(env)
+func createTarredFileOf(value string, config types.FileReferenceConfig) (bytes.Buffer, error) {
+	mode, uid, gid, err := makeTarFileEntryParams(config)
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed parsing target file parameters")
+	}
+
 	b := bytes.Buffer{}
 	tarWriter := tar.NewWriter(&b)
-	mode := uint32(0o444)
+	valueAsBytes := []byte(value)
+	header := &tar.Header{
+		Name:    config.Target,
+		Size:    int64(len(valueAsBytes)),
+		Mode:    mode,
+		ModTime: time.Now(),
+		Uid:     uid,
+		Gid:     gid,
+	}
+	err = tarWriter.WriteHeader(header)
+	if err != nil {
+		return bytes.Buffer{}, err
+	}
+	_, err = tarWriter.Write(valueAsBytes)
+	if err != nil {
+		return bytes.Buffer{}, err
+	}
+	err = tarWriter.Close()
+	return b, err
+}
+
+func createTarArchiveOf(path string, config types.FileReferenceConfig) (bytes.Buffer, error) {
+	// need to treat files and directories differently
+	fi, err := os.Stat(path)
+	if err != nil {
+		return bytes.Buffer{}, err
+	}
+
+	// if path is not directory, try to treat it as a file by reading its value
+	if !fi.IsDir() {
+		buf, err := os.ReadFile(path)
+		if err == nil {
+			return createTarredFileOf(string(buf), config)
+		}
+	}
+
+	mode, uid, gid, err := makeTarFileEntryParams(config)
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed parsing target file parameters")
+	}
+
+	subdir := os.DirFS(path)
+	b := bytes.Buffer{}
+	tarWriter := tar.NewWriter(&b)
+
+	// build the tar by walking instead of using archive/tar.Writer.AddFS to be able to adjust mode, gid and uid
+	err = fs.WalkDir(subdir, ".", func(filePath string, d fs.DirEntry, err error) error {
+		header := &tar.Header{
+			Name:    filepath.Join(config.Target, filePath),
+			Mode:    mode,
```

Comment:
:question: Should directories have the same mode as files? Or should they perhaps have the exec bit set for the owner/group/other to access them?